crates/oxc_linter/src/rules/promise/no_nesting.rs

```diff
@@ -0,0 +1,298 @@
+    style,
+    pending
+);
+
+fn is_inside_promise(node: &AstNode, ctx: &LintContext) -> bool {
+    if !matches!(node.kind(), AstKind::Function(_) | AstKind::ArrowFunctionExpression(_))
+        || !matches!(ctx.nodes().parent_kind(node.id()), Some(AstKind::Argument(_)))
+    {
+        return false;
+    }
+
+    ctx.nodes()
+        .ancestors(node.id())
+        .nth(2)
+        .is_some_and(|node| node.kind().as_call_expression().is_some_and(has_promise_callback))
+}
+
+/// Gets the closest promise callback function of the nested promise.
+fn closest_promise_cb<'a, 'b>(
+    node: &'a AstNode<'b>,
+    ctx: &'a LintContext<'b>,
+) -> Option<&'a CallExpression<'b>> {
+    ctx.nodes()
+        .ancestors(node.id())
+        .filter_map(|node| node.kind().as_call_expression())
+        .filter(|a| has_promise_callback(a))
+        .nth(1)
+}
+
+fn has_promise_callback(call_expr: &CallExpression) -> bool {
+    matches!(
+        call_expr.callee.as_member_expression().and_then(MemberExpression::static_property_name),
+        Some("then" | "catch")
+    )
+}
+
+fn is_promise_then_or_catch(call_expr: &CallExpression) -> Option<String> {
+    let member_expr = call_expr.callee.get_member_expr()?;
+    let prop_name = member_expr.static_property_name()?;
+
+    // For example: hello.then(), hello.catch()
+    if matches!(prop_name, "then" | "catch") {
+        return Some(prop_name.into());
+    }
+
+    None
+}
+
+/// Checks if we can safely unnest the promise callback.
+///
+/// 1. This function gets variable bindings defined in closest parent promise callback function
+///    scope.
+///
+/// 2. Checks if the argument callback of the nested promise call uses any of these variables
+///    and if so returns `false` to denote that the promises cannot be safely unnested.
+///
+/// Here is an example of a nested promise which isn't safe to nest without further refactoring.
+/// ```javascript
+/// doThing()
+///  .then(a => getB(a) <---- 1. Get this scopes bound variables
+///    .then(b => getC(a, b)) <--- 2. Check for references to the bound variables from 1.
+///  )
+/// ```
+///
+/// In this case unnesting is not safe as doing so would result in `a` being undefined in the
+/// expression `getC(a, b)`, as seen below in the unnested version of the example:
+/// ```javascript
+/// doThing()
+///  .then(a => getB(a))
+///  .then(b => getC(a, b))
+/// ```
+fn can_safely_unnest(
+    call_expr: &CallExpression,
+    closest: &CallExpression,
+    ctx: &LintContext,
+) -> bool {
+    let mut closest_cb_scope_bindings = vec![];
+
+    closest.arguments.iter().for_each(|new_expr| {
+        let Some(arg_expr) = new_expr.as_expression() else {
+            return;
+        };
+
+        match arg_expr {
+            Expression::ArrowFunctionExpression(arrow_expr) => {
+                let func_scope = arrow_expr.scope_id();
+                closest_cb_scope_bindings =
+                    ctx.scopes().get_bindings(func_scope).iter().collect_vec();
```

Comment:
i think we want to extend the vec not replace it?

Reply:
i maybe wrong though?

Reply:
mayube the best sokutions is something like:
1. make `closest_cb_scope_bindings` a set - we only use the value of the map
2. extend it in here? (`get_bindings().iter().for_each(|id| closes....add(id))`

Reply:
The current implementation relies on just using those bindings to see if they are used as references in another span. I have pushed up a small refactor which should hopefully make the intent clearer.

Let me know what you think.

Reply:
yes this looks much better